### ISSUE TITLE
FC precision: 4 false positives / 91.1% precision on the 303-sample test set (both AST and torch-free paths — a MODEL-quality matter, not a D27 regression)

### ISSUE THREAD

**Issue by syamaner** [syamaner/coffee-first-crack-detection#55]
Surfaced by the D27 Phase 1 torch-free eval (#54). On the **current 303-sample** test split (`data/splits/test/`, 261 negatives / ~42 positives), the ONNX int8 model scores **98.3% accuracy / 91.1% precision / 4 false positives** — and it scores **identically on the old `ASTFeatureExtractor` path and the new numpy/scipy `MelFrontend` path** (numerically equivalent, <1.3e-05). So this is the **model**, not the front-end swap — filing it separately so it never holds D27 hostage.

## Why it matters
4 false positives = 4 spurious "first crack detected" events. On a real roast a false FC would start the development-% clock early and mislead the advisor/operator — so precision matters for the appliance, more than raw accuracy.

## Context
- The recorded **191-sample** v2 baseline was 96.86% acc / 96.9% precision / ≤1 FP. The current split is **larger and harder** (more negatives), which moved the precision/FP numbers for BOTH paths equally. The 96.9%/≤1FP "gate" in #54 was written against the 191-sample set; D27 is regression-free against it (equivalence proven).

## Investigate / levers
- Which 4 samples false-positive, and what audio traits do they share (transients, mic clipping, background)?
- Confidence threshold: the Pi profile already uses `confidence_threshold=0.90` + a **≥3-positive-in-30s confirmation** window — does that filter these 4 at the live-detection layer even if the per-window eval counts them?
- More negatives in training (the #176/#181 full-roast recordings grow the negative class), and/or confirmation-window tuning.

Non-blocking for D27 Phase 1 (accepted, regression-free). Refs #54, roastpilot-plan D27.

**Comment by syamaner** [comment]
## Analysis (7 Jul, read-only) — the headline number isn't reproducible from committed data; one sweep pins it

**First finding: the exact "4 FP / 91.1%" is not in any committed artifact.** It appears only in prose in the merged PR #56 body (which explicitly deferred it here as a separate model-quality matter). The equivalence test PR #56 leans on (`tests/test_mel_frontend.py`) iterates all 303 WAVs but only asserts a feature max-abs-diff bound — it never classifies, so the FP list can't be mined from it. The saved 303-set confusion matrices disagree with the claim *and* each other:

| Source | TN/FP/FN/TP | Precision | Recall |
|---|---|---|---|
| baseline_v5 (AST, production) — `experiments/baseline_v5/evaluation/test_results.json` | 255/**6**/1/41 | 87.2 % | 97.6 % |
| baseline_v6 (reverted) — `experiments/baseline_v6/evaluation/test_results.json` | 256/**5**/3/39 | 88.6 % | 92.9 % |
| Issue/PR #56 claim (unsaved) | ?/**4**/?/? | 91.1 % | not stated |

The 4-FP figure most plausibly reflects the **ONNX-INT8** model (quantization shifts a few borderline scores vs the raw checkpoint) — but that's a hypothesis until the run below.

**Pinning the 4 FPs is one local forward pass** (model `exports/onnx/int8/model_quantized.onnx` + data `data/splits/test/` are present; no training, no release):
```
python scripts/evaluate_onnx.py --onnx-dir exports/onnx/int8 --test-dir data/splits/test \
  --threshold-sweep --output results/v3_303_threshold_sweep.json
```
then filter `label_id==0 & prob>=threshold` for the filenames. Only `evaluate_onnx.py --threshold-sweep` persists per-sample predictions; `evaluate.py` (AST) does not.

**Ranked levers (cheapest first):**
1. **Threshold + confirmation-window gate (no retrain).** `pi_inference` already runs `threshold 0.90` + a 3-of-4-window / 30 s confirmation gate — designed to suppress exactly these borderline FPs. The only committed sweep (old **45-sample** set) showed 0.6→0.90 cut FP 2→1 at recall 95.5→90.9 %; **no 303-set sweep exists**, so run the command above to confirm the 4 scores fall under the gate. Near-zero cost if they do.
2. **Hard-negative data addition** — the project's own "highest-impact" lever (`docs/hyperparameter_tuning.md`); a real training run, release-gated, out of scope.
3. Model/labelling — least promising (no mislabel evidence; v3 full-fine-tune and v6 more-regularized both did *worse*).
4. Accept-as-is — plausible but can't be certified (the recall/FN for the exact 4-FP run isn't saved).

**Operating point is right:** the repo prioritizes recall (a missed FC over-roasts until an operator override; a spurious FC only trims heat slightly early — bounded, recoverable), so a recall-favoring model + a precision-favoring runtime gate is the correct posture. The one unverified piece is whether that gate catches *these* 4.

**Also:** the epic's "Latest Results" table (`docs/state/epics/coffee-first-crack-detection.md`) is **stale** (still 45-sample ONNX numbers, no v6 / 303-set ONNX row) — worth refreshing when the sweep runs.

**Recommended next step:** run the one sweep command to establish which model/config the 4-FP number refers to and whether the 0.90 gate already suppresses them — that decides between lever 1 (near-free) and lever 2 (retrain). (Analysis was read-only; the sweep itself is a cheap local run I can do on request.)

**Comment by syamaner** [comment]
## Reconciled — two legitimate numbers, one stale artifact (12 Jul)

Ran the sweep this thread's analysis pointed to. Confirmed root cause and pinned both real numbers.

**Root cause:** `experiments/baseline_v5/evaluation/test_results.json` (the source of the README's "1 FN, 6 FP") was written at 21:43 on 12 Apr; `data/splits/test/` was regenerated at 21:56 the same session — 13 minutes later, same 21 recordings / same seed=42 / same 42-FC/261-NFC composition (confirmed via `split_report.md`, no data added — the later "el-salvador-diamante" raw recordings are from 31 May and aren't in this split at all), but the chunked WAVs were rewritten. That committed JSON scored a since-replaced version of the split. It's a stale local artifact (`experiments/` is `.gitignore`'d, so it never got fixed by a later PR) — regenerated it locally; also updating `README.md`, `results/README.md`, and the epic status table, all previously transcribed from that stale file.

**Fresh, reproducible numbers against `experiments/baseline_v5/checkpoint-best` and the current `data/splits/test/` (deterministic, `crop_mode="center"`, reproduced twice each):**

| Config | Accuracy | Precision | Recall | F1 | ROC-AUC | Confusion (TN/FP/FN/TP) |
|---|---|---|---|---|---|---|
| AST / PyTorch fp32 | 98.02% | 89.13% | 97.62% | 0.932 | 0.9979 | 256/5/1/41 |
| ONNX fp32 | 98.02% | 89.13% | 97.62% | 0.932 | 0.9979 | 256/5/1/41 (exact match to AST — expected) |
| **ONNX int8** | **98.35%** | **91.11%** | 97.62% | 0.943 | 0.9976 | **257/4/1/41** |

**This issue's "4 FP / 91.1% precision" is real and now reproduced exactly — it's the ONNX INT8 export's number**, not the AST/PyTorch checkpoint's. The checkpoint result you'd get by running `coffee_first_crack.evaluate` is 5 FP (not the 6 in the old README, not the 4 here) — a third, previously-unrecorded number, now the current committed one.

Net effect: INT8 quantization did **not** cost precision on this 303-set — it came out marginally *ahead* of fp32 (4 FP vs 5, +1.98pp precision). That's a quantization-noise-sized shift on one run, not a general guarantee, but it means lever 1 from the earlier analysis (does the runtime confirmation gate already catch these?) is now less urgent — the INT8 model as-deployed already has one fewer FP than the fp32/AST reference score the README used to advertise.

**Reproduce:**
```bash
python -m coffee_first_crack.evaluate --model-dir experiments/baseline_v5/checkpoint-best --test-dir data/splits/test
python scripts/evaluate_onnx.py --onnx-dir exports/onnx/fp32 --test-dir data/splits/test
python scripts/evaluate_onnx.py --onnx-dir exports/onnx/int8 --test-dir data/splits/test
```

Raw JSON + latency benchmark: `results/baseline_v5_303set/`.

**Status:** the "which number is right" question this issue opened with is answered — closing that part out. Leaving this issue **open** for the still-outstanding investigation items (which specific samples FP, audio traits, whether the 0.90 + confirmation-window gate suppresses them) since those weren't run here — that's the 4 (now confirmed real) ONNX-int8 false positives, not a new sweep target.
